File: erlang/etudes/ch5/ask_area.erl
-module(ask_area).
-export([area/0]).

%% @doc Calculates area of various shapes.

area() ->
    Answer = strip_newline(io:get_line("R)ectangle, T)riangle, or E)llipse > ")),
    Shape = char_to_shape(hd(Answer)),
    {F, S} = case Shape of
        rectangle -> get_dimensions("width", "height");
        triangle -> get_dimensions("base", "height");
        ellipse -> get_dimensions("major axis", "minor axis");
        unknown -> { error, "Unknown shape x."}
    end,
    calculate(Shape, F, S).

calculate(unknown, _, Msg) -> io:format("~s~n", [Msg]);
calculate(_, error, _) -> io:format("Error in first number.~n");
calculate(_, _, error) -> io:format("Error in second number.~n");
calculate(_, A, B) when A < 0; B < 0 -> io:format("Both numbers must be greater than or equal to zero.~n");
calculate(Shape, A, B) -> io:format("~w~n", [geom:area({Shape, A, B})]).

char_to_shape(Char) ->
    case string:to_upper(Char) of
        $R -> rectangle;
        $T -> triangle;
        $E -> ellipse;
        _ -> unknown
    end.

strip_newline(String) -> string:strip(String,right,$\n).

get_number(Prompt) ->
    Answer = strip_newline(io:get_line("Enter " ++ Prompt ++ " > ")),
    case string:to_float(Answer) of
        {error, _} -> string:to_integer(Answer);
        Result -> Result
    end.

get_dimensions(First, Second) ->
    { N, _ } = get_number(First),
    { M, _ } = get_number(Second),
    {N, M}.

File: elixer/etudes/ch05/ask_area.ex
defmodule AskArea do
  def area() do
    ans = IO.gets("R)ectangle, T)riangle, or E)llipse: ")
    shape = char_to_shape(String.upcase(String.first(ans)))
    {a, b} = get_dimensions(shape)
    cond do
      a == :error or b == :error -> IO.puts("Both entries must be numbers.")
      a < 0 or b < 0 -> IO.puts("Both numbers must be greater than or equal to zero.")
      shape == :unknown -> IO.puts("Unknown shape #{ans}")
      true -> Geom.area(shape, a, b)
    end
  end

  defp char_to_shape(char) when char == "R", do: :rectangle
  defp char_to_shape(char) when char == "T", do: :triangle
  defp char_to_shape(char) when char == "E", do: :ellipse
  defp char_to_shape(_char), do: :unknown

  defp get_number(prompt) do
    ans = String.strip(IO.gets("Enter #{prompt} > "))
    cond do
      Regex.match?(~r/\d+.\d+/, ans) -> String.to_float(ans)
      Regex.match?(~r/\d+/, ans) -> String.to_integer(ans)
      true -> :error
    end
  end
  defp get_dimensions(prompt1, prompt2), do: { get_number(prompt1), get_number(prompt2) }
  defp get_dimensions(shape) when shape == :rectangle, do: get_dimensions("width", "height")
  defp get_dimensions(shape) when shape == :triangle, do: get_dimensions("base", "height")
  defp get_dimensions(shape) when shape == :ellipse, do: get_dimensions("major radius", "minor radius")
  defp get_dimensions(shape) when shape == :unknown, do: { 0, 0 }
end
